### PR TITLE
[V3] [Audio] queues, mixing, and downloader

### PR DESCRIPTION
**Pull request by flapjax**
I'm going to get this massive boulder rolling... here are a few features I've implemented.

- basic queue system for downloaded songs (use `[p]play`)
- simultaneous/overlayed playing of sounds (nice proof of concept for sfx, use `[p]local`)
- data folders for local sounds and the download cache (using `cog_data_path` of course!)

### CODE REVIEW

File: cogs/audio/audio.py [hunk 1]
```diff
@@ -1,112 +1,367 @@
...
         self.bot = bot
+        self.conf = Config.get_conf(self, identifier=8675309)
+        self.conf.register_global(
+            **self.default_global_settings
+        )
+        self.conf.register_guild(
+            **self.default_guild_settings
+        )
+        self.data_path = cog_data_path(self)
+        self.local_path = os.path.join(self.data_path, 'local')
```
Comment (tekulvw) [inline cogs/audio/audio.py]: take a look at the python documentation for `pathlib`, it'll make your life easier by turning this `os.path` crap into `self.data_path / 'local'`

File: cogs/audio/audio.py [hunk 2]
```diff
@@ -1,112 +1,367 @@
...
+        self.downloader_path = os.path.join(self.data_path, 'cache')
+        self.downloader = Downloader(self.downloader_path)
+        if not os.path.exists(self.local_path):
+            os.makedirs(self.local_path)
```
Comment (tekulvw) [inline cogs/audio/audio.py]: `pathlib` will also make this easier, turning it into `self.local_path.mkdir(exists_ok=True)`

Comment (flapjax) [reply cogs/audio/audio.py]: Done and done! Only problem I ran into with pathlib was that FFmpegPCMAudio did not enjoy being fed the Path object. Using the string representation worked.

**Comment by tekulvw** [comment]
@flapjax do you want to see this merged or do you want to hold off for a bit?

**Comment by flapjax** [comment]
Hold off for now, unless someone else has tested it recently. Imports are likely broken since I opened this before we changed up the folder structure a while back.

**Comment by tekulvw** [comment]
Closing because of staleness and further work on alternative audio solutions.

**Pull request closed without merging**